corehq/apps/enterprise/api/resources.py

```diff
@@ -138,12 +144,64 @@ def convert_datetime(cls, datetime_string):
         return time.isoformat()
 
 
-class ODataEnterpriseResource(ODataResource):
+class ODataEnterpriseReportResource(ODataResource):
+    REPORT_SLUG = None  # Override with correct slug
+    # If this delay is too quick, clients like PowerBI
+    # will hit their maximum number of retries before the report is ever generated.
+    # If the delay is too long, then even reports that would generate in well under the retry
+    # window will be subject to this delay (PowerBI will subject them to it twice,
+    # as the data preview and actual request perform separate queries
+    RETRY_IN_PROGRESS_DELAY = 60
+    RETRY_CONFLICT_DELAY = 120
+
     class Meta(ODataResource.Meta):
         authentication = EnterpriseODataAuthentication()
 
+    def get_object_list(self, request):
+        query_id = request.GET.get('query_id', None)
+        progress = ReportTaskProgress(
+            self.REPORT_SLUG, request.couch_user.username, query_id=query_id)
+        status = progress.get_status()
+        if status == ReportTaskProgress.STATUS_COMPLETE:
+            try:
+                # ensure this is for the same parameters
+                if progress.get_task() != self.get_report_task(request):
+                    raise ImmediateHttpResponse(
+                        response=http.HttpTooManyRequests(headers={'Retry-After': self.RETRY_CONFLICT_DELAY}))
```

Comment:
Does this need to be in the `try` block? I think it does not.

Reply:
Good call, moved outside of the block